src/libexpr/flake/flake.cc

```diff
@@ -793,6 +793,104 @@ static RegisterPrimOp r2({
     .experimentalFeature = Xp::Flakes,
 });
 
+static void prim_parseFlakeRef(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    std::string flakeRefS(state.forceStringNoCtx(
+        *args[0], pos,
+        "while evaluating the argument passed to builtins.parseFlakeRef"));
+    auto attrs = parseFlakeRef(flakeRefS, {}, true).toAttrs();
+    auto binds = state.buildBindings(attrs.size());
+    for ( const auto & [key, value] : attrs ) {
+        Symbol  s  = state.symbols.create(key);
+        Value & vv = binds.alloc(s);
+        if (std::holds_alternative<std::string>(value)) {
+            vv.mkString(std::get<std::string>(value));
+        }
+        else if (std::holds_alternative<uint64_t>(value)) {
+            vv.mkInt(std::get<uint64_t>(value));
+        }
+        else { /* bool */
+            vv.mkBool(std::get<Explicit<bool>>(value).t);
+        }
+
+    }
+    v.mkAttrs(binds);
+}
+
+static RegisterPrimOp r3({
+    .name =  "__parseFlakeRef",
+    .args = {"flake-ref"},
+    .doc = R"(
+      Parse a flake reference, and return its exploded form. For example:
+
+      ```nix
+      builtins.parseFlakeRef "github:NixOS/nixpkgs/23.05?dir=lib"
+      # ==> { dir = "lib"; owner = "NixOS"; ref = "23.05"; repo = "nixpkgs";
+      #       type = "github"; }
+      ```
+    )",
+    .fun = prim_parseFlakeRef,
+    .experimentalFeature = Xp::Flakes,
+});
+
+
+static void prim_flakeRefToString(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    state.forceAttrs(*args[0], noPos, "while evaluating the argument passed to builtins.flakeRefToString");
+    fetchers::Attrs attrs;
+    for (const auto & attr : *args[0]->attrs) {
+        switch (attr.value->type())
+        {
+            case nInt:
+                attrs.emplace(state.symbols[attr.name],
+                              (uint64_t) attr.value->integer);
+                break;
+            case nBool:
+                attrs.emplace(state.symbols[attr.name],
+                              Explicit<bool> { attr.value->boolean });
+                break;
+            case nString:
+                attrs.emplace(state.symbols[attr.name],
+                              std::string(attr.value->str()));
+                break;
+            case nThunk:
+            case nFloat:
+            case nPath:
+            case nNull:
+            case nAttrs:
+            case nList:
+            case nFunction:
+            case nExternal:
+            default:
+                state.error(
```

Comment:
This error should be impossible right? So probably this should be an assertion failure.